accepted/2023/owner-author-pmui.md

```diff
@@ -0,0 +1,78 @@
+# ***Display Trusted Owners in PM UI***
+<!-- Replace `Title` with an appropriate title for your design -->
+
+- Donnie Goodson ([donnie-msft](https://github.com/donnie-msft))
+- GitHub Issue https://github.com/NuGet/Home/issues/12501
+
+## Summary
+
+When available, highlight in the Packages Manager UI (PM UI) that the owner is trusted by the package source with a NuGet package icon. 
+In the case of `nuget.org`, also provide a hyperlink to the owner's profile page (ie, `https://www.nuget.org/profiles/<owner name>`).
+A Trusted Owner is defined as an owner that's indicated by a package source's [Search API](https://learn.microsoft.com/nuget/api/search-query-service-resource#search-result), as opposed to the field in `.nuspec`.
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+
+Customers need a clear understanding of who created a NuGet package in order to trust that package.
+The term "owner" has become overloaded as NuGet has evolved, which has caused confusion. 
+The "author" metadata is also misleading as it's a free-form text field which doesn't imply any credentials.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+![Owners from NuGet.org are shown beside each package ID with a hyperlink and a NuGet Package Icon](../../meta/resources/OwnerAuthor/ownerNuGetOrgLinks.png)
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+Currently, `Owners` JSON parsed from the Search API is immediately converted into a comma-delimited string of owners.
+To avoid breaking any clients depending on this, `Owners` will continue being a pre-processed string with comma-delimited owners.
+A new property, `OwnersEnumerable` (naming TBD) will be added to [IPackageSearchMetadata](https://github.com/NuGet/NuGet.Client/blob/dc2fadacd311646754ce6b062489b403338972aa/src/NuGet.Core/NuGet.Protocol/Model/IPackageSearchMetadata.cs) and [PackageSearchMetadataContextInfo](https://github.com/NuGet/NuGet.Client/blob/dc2fadacd311646754ce6b062489b403338972aa/src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/ContextInfos/PackageSearchMetadataContextInfo.cs).
+The enumerable owners will allow displaying custom UI for each owner.
```

Comment:
What happens if there are no owners? What is displayed?

```kusto
NiPackageOwners
| where array_length(Owners) == 0
| join kind=inner NiPackageDownloads on LowerId
| join kind=inner NiPackageVersions on Identity
| where IsListed == true
| distinct ['Package ID'] = Id, TotalDownloads
| order by TotalDownloads desc
| take 5
```

|Package ID|TotalDownloads|
|---|---|
|NDesk.Options|3158883|
|Core.Common|404489|
|CommonServiceLocator.WindsorAdapter|374352|
|Mono.Posix-4.5|281608|
|Core.Calendar|264348|

Reply:
Author will be shown as the fallback

Reply:
This may be fine, if it's not linked as an owner. But this could confuse or even deceive our users. Consider these packages that have no owners but have authors that match real NuGet.org usernames:

```kusto
let LatestVersionWithNoOwners = NiPackageOwners
| where array_length(Owners) == 0
| join kind=inner NiPackageDownloads on LowerId
| join kind=inner NiPackageVersions on Identity
| where IsLatestSemVer2 == true
| join kind=inner NiPackageManifests on Identity
| distinct ['Package ID'] = Id, LatestVersion = Version, TotalDownloads, Authors
| order by TotalDownloads desc; 
let AllOwners = NiPackageOwners
| mv-expand Owner = Owners to typeof(string)
| distinct Owner, LowerOwner = tolower(Owner);
LatestVersionWithNoOwners
| extend LowerAuthors = tolower(trim("\\s", Authors))
| join kind=leftouter AllOwners on $left.LowerAuthors == $right.LowerOwner
| project ['Package ID'], TotalDownloads, Authors, ['Matching owner'] = Owner
| where isnotempty(['Matching owner'])
| order by TotalDownloads
| take 10
```

|Package ID|TotalDownloads|Authors|Matching owner|
|---|---|---|---|
|Mvc3CodeTemplatesCSharp|31125|Microsoft|Microsoft|
|OData.Helper|26299|Microsoft|Microsoft|
|WindowsAzureStorage.Helper|20959|Microsoft|Microsoft|
|SharedCode.Common.Security|9604|e-Construmarket|e-Construmarket|
|EFCTP4|7070|Microsoft|Microsoft|
|Mvc3CodeTemplatesVB|5691|Microsoft|Microsoft|
|FsLib.Common|5577|ifish|iFish|
|EFCTP4.SampleBlogModel|4199|Microsoft|Microsoft|
|Moosend.Wrappers.CSharpWrapper|4059|moosend|moosend|
|Lykke.Service.LykkeDevelopers.Client|3197|Lykke|lykke|

Reply:
I'd only render owner hyperlinks when we have owner metadata from the package source. If the author looks like an owner, it won't have a hyperlink (or indicative icon, if we keep that). This is the reason my alternative introduced a new term, 'contributors', to start differentiating these with a descriptor.

Reply:
I think showing either owners or authors but not both is a better design, but I don't feel very strongly. I think I am erring on the side of caution (authors are not secured like owners) and you are erring on the side on usability (provide the user with what information we have). Pros and cons. Perhaps we can raise the point to others for additional feedback? I'll leave the decision to you.

Reply:
> I think showing either owners or authors but not both (per source) is a better design

👍and that's what I'm proposing as well. There is never both owner/author shown. If we have owner, show it with links, if we don't, show author which will never have links.

Reply:
I didn't realize you were writing about the scenario of merging metadata when the `All` source is selected. 

For Case 1 & 2, not showing anything (even author) is an interesting idea. My instinct is that some customers rely on author to quickly spot which internal team created a package. Removing it entirely is enticing, but I think I'll avoid that for this spec.

My plan was to show Trusted Owner provided from a source, if available, otherwise continue showing author. 

The notion of an icon paired with each Owner plays an important role in the UI. We need a fixture that always means Trusted Owner, even if there isn't a profile hyperlink that can be attached to the owner.

> Said another way, do you want "Microsoft" (unlinked, the authors value) to show when you search for OData.Helper on nuget.org?

I'm leaning yes, since it already does so. The new affordances won't be there, so we can start helping customers learn that this rendering means something lesser than a Trusted Owner.

However, I like how nuget.org has an explanation when there's no owners, which I could implement in the Tooltip of the packages list (and maybe another icon/affordance):  `This package has no owners and is not being actively maintained.`.

Perhaps these can be done in iterations (1. keep author, introduce new UI affordances. 2. hide author in certain cases, 3. hide author from packages list in all cases)....

Reply:
I've scoped this spec to single-source only, and mentioned this in Future Possibilities. Leaving the comment open for now so others are more likely to see it.

Reply:
@donnie-msft , I was about to start a similar thread when I saw this conversation. First of all, thank you and @joelverhagen for the in-depth discussion of this edge case.

> Anyhow, if nuget.org is the only selected feed I would expect authors to appear on no results since displaying them for ownerless packages right next to results with owners displayed might accidentally lend trustworthiness to the authors value that really should be reserved for the owners value.

I share this concern. The 'author' metadata can be easily gamed by attackers, so we shouldn't display it for nuget.org packages. Even if it looks differently than 'Trusted owners', I don't know if users will be able to easily tell the difference and make the right trust decision. 
Please don't consider my comment as blocking. Happy to have an offline chat if it helps. :)

Reply:
I've redacted the fallback to author when the package source provides Trusted Owners.
https://github.com/NuGet/Home/blob/dev-donnie-msft-ownerAuthorSpec/accepted/2023/owner-author-pmui.md#packages-without-a-trusted-owner